src/i_oalsound.c

```diff
@@ -0,0 +1,662 @@
+{
+    alcSuspendContext(alcGetCurrentContext());
+}
+
+static void AL_APIENTRY wrap_ProcessUpdatesSOFT(void)
+{
+    alcProcessContext(alcGetCurrentContext());
+}
+
+static void InitDeferred(void)
+{
+    // Deferred updates allow values to be updated simultaneously.
+    // https://openal-soft.org/openal-extensions/SOFT_deferred_updates.txt
+
+    if (alIsExtensionPresent("AL_SOFT_deferred_updates") == AL_TRUE)
+    {
+        alDeferUpdatesSOFT = FUNCTION_CAST(LPALDEFERUPDATESSOFT,
+                                           alGetProcAddress("alDeferUpdatesSOFT"));
+        alProcessUpdatesSOFT = FUNCTION_CAST(LPALPROCESSUPDATESSOFT,
+                                             alGetProcAddress("alProcessUpdatesSOFT"));
+
+        if (alDeferUpdatesSOFT && alProcessUpdatesSOFT)
+        {
+            return;
+        }
+    }
+
+    alDeferUpdatesSOFT = FUNCTION_CAST(LPALDEFERUPDATESSOFT, &wrap_DeferUpdatesSOFT);
+    alProcessUpdatesSOFT = FUNCTION_CAST(LPALPROCESSUPDATESSOFT, &wrap_ProcessUpdatesSOFT);
+}
+
+void I_OAL_ShutdownSound(void)
+{
+    if (!oal)
+    {
+        return;
+    }
+
+    if (oal->sources)
+    {
+        alDeleteSources(MAX_CHANNELS, oal->sources);
+        free(oal->sources);
+    }
+
+    if (oal->active)
+    {
+        free(oal->active);
+    }
+
+    if (oal->buffers)
+    {
+        if (oal->num_buffers > 0)
+        {
+            alDeleteBuffers(oal->num_buffers, oal->buffers);
+        }
+        free(oal->buffers);
+    }
+
+    alcMakeContextCurrent(NULL);
+    if (oal->context)
+    {
+        alcDestroyContext(oal->context);
+    }
+
+    if (oal->device)
+    {
+        alcCloseDevice(oal->device);
+    }
+
+    free(oal);
+    oal = NULL;
+}
+
+static void SetResampler(ALuint *sources)
+{
+    const char *resampler_name = oal_resamplers[snd_resampler];
+    LPALGETSTRINGISOFT alGetStringiSOFT = NULL;
+    ALint i, num_resamplers, def_resampler;
+
+    if (alIsExtensionPresent("AL_SOFT_source_resampler") != AL_TRUE)
+    {
+        printf(" Resampler info not available!\n");
+        return;
+    }
+
+    alGetStringiSOFT = FUNCTION_CAST(LPALGETSTRINGISOFT,
+                                     alGetProcAddress("alGetStringiSOFT"));
+
+    if (!alGetStringiSOFT)
+    {
+        fprintf(stderr, " alGetStringiSOFT() is not available.\n");
+        return;
+    }
+
+    num_resamplers = alGetInteger(AL_NUM_RESAMPLERS_SOFT);
+    def_resampler = alGetInteger(AL_DEFAULT_RESAMPLER_SOFT);
+
+    if (!num_resamplers)
+    {
+        printf(" No resamplers found!\n");
+        return;
+    }
+
+    for (i = 0; i < num_resamplers; i++)
+    {
+        if (!strcasecmp(resampler_name, alGetStringiSOFT(AL_RESAMPLER_NAME_SOFT, i)))
+        {
+            def_resampler = i;
+            break;
+        }
+    }
+    if (i == num_resamplers)
+    {
+        printf(" Failed to find resampler: '%s'.\n", resampler_name);
+        return;
+    }
+
+    for (i = 0; i < MAX_CHANNELS; i++)
+    {
```

Comment:
Do we have to re-cache the SFX when we change the resampling method? I guess not, but better ask now...

Reply:
No, the resampler method is applied at the mixing phase and not during the creation of the buffer.